src/AI/JakeMan.java

```diff
@@ -0,0 +1,770 @@
+    log(String.format("  %s is traveling toward %s at %s via %s  mustMove?: %s",
+                          unit.toStringWithLocation(),
+                          gameMap.getLocation(goal).getEnvironment().terrainType, goal,
+                          pathPoint, mustMove));
+    for( XYCoord xyc : destinations )
+    {
+      log(String.format("    is it safe to go to %s?", xyc));
+      if( !isDudeFree(gameMap, unit, xyc, false) )
+        continue;
+
+      GameAction action = null;
+      Unit resident = gameMap.getLocation(xyc).getResident();
+      if( null != resident && unit != resident )
+      {
+        if( unit.CO == resident.CO && !resident.isTurnOver )
+          action = evictUnit(gameMap, unit, resident, avoidProduction);
+        if( null != action ) return action;
+        continue;
+      }
+      log(String.format("    Yes"));
+
+      GamePath movePath = Utils.findShortestPath(unit, xyc, gameMap);
+      ArrayList<GameActionSet> actionSets = unit.getPossibleActions(gameMap, movePath, ignoreResident);
+      if( actionSets.size() > 0 )
+      {
+        // Since we're moving anyway, might as well try shooting the scenery
+        for( GameActionSet actionSet : actionSets )
+        {
+          if( actionSet.getSelected().getType() == UnitActionFactory.ATTACK )
+          {
+            action = findBestAttack(gameMap, unit, actionSet.getGameActions());
+          }
+        }
+
+        if( null == action && movePath.getPathLength() > 1) // Just wait if we can't do anything cool
+          action = new WaitLifecycle.WaitAction(unit, movePath);
+        return action;
+      }
+    }
+    return null;
+  }
+
+  public static GameAction findBestAttack(GameMap gameMap, Unit unit, ArrayList<GameAction> actionSet)
+  {
+    double bestDamage = 0;
+    GameAction bestAttack = null;
+    for( GameAction attack : actionSet )
+    {
+      double damageValue = AICombatUtils.scoreAttackAction(unit, attack, gameMap,
+          (results) -> {
+            final Unit defender = results.defender.unit;
+            int loss   = Math.min(unit    .getHP(), (int)results.attacker.getPreciseHPDamage());
+            int damage = Math.min(defender.getHP(), (int)results.defender.getPreciseHPDamage());
+
+            // Convert to abstract value
+            int extraLoss = 0;
+            if( loss >= 1 && unit.getHP() == UnitModel.MAXIMUM_HP )
+              extraLoss += STAY_UNHURT_BIAS;
+            if( loss >= unit.getHP() )
+              extraLoss += STAY_ALIVE_BIAS;
+            loss *= unit.getCost();
+            loss += extraLoss;
+            damage *= defender.getCost();
+            if( isThreatenedBy(unit.model, defender.model) )
+              damage *= FIRSTSTRIKE_ON_THREAT_WEIGHT;
+            // Value damage to hurt units less
+            damage *= defender.getHPFactor();
+
+            return (double)(damage - loss);
+          }, (terrain, params) -> 0.01); // Attack terrain, but don't prioritize it over units
```

Comment:
This will cause Jakeman to allocate infantry to shoot terrain (forever), when they could be doing more useful things.